Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs

```diff
@@ -0,0 +1,210 @@
+                sprite.Color = contents.SubstanceColor;
+        }
+
+        private void SpreadSmoke()
+        {
+            if (_amount == 0)
+            {
+                return;
+            }
+
+            if (!Owner.TryGetComponent(out SnapGridComponent snapGrid))
+                return;
+
+            void SpreadToDir(Direction dir)
+            {
+                foreach (var neighbor in snapGrid.GetInDir(dir))
+                {
+                    if (neighbor.TryGetComponent(out SmokeComponent comp) && comp._inception == _inception)
+                        return;
+
+                    if (neighbor.TryGetComponent(out AirtightComponent airtight) && airtight.AirBlocked)
+                        return;
+                }
+                var newSmoke = Owner.EntityManager.SpawnEntity("smoke_cloud", snapGrid.DirectionToGrid(dir));
+                var smokeComponent = newSmoke.GetComponent<SmokeComponent>();
+
+                if (Owner.TryGetComponent(out SolutionContainerComponent contents))
+                {
+                    var solution = contents.Solution.Clone();
+                    smokeComponent.TryAddSolution(solution);
+                }
+                smokeComponent.Start(_inception,_amount-1,_lifetime-_timer-_spreadDelay, _spreadDelay, _averageLifetime);
+            }
+
+            SpreadToDir(Direction.North);
+            SpreadToDir(Direction.East);
+            SpreadToDir(Direction.South);
+            SpreadToDir(Direction.West);
+        }
+
+        private void ReactWithTileAndEntities()
+        {
+            if (!Owner.TryGetComponent(out SolutionContainerComponent contents))
+                return;
+
+            var mapGrid = _mapManager.GetGrid(Owner.Transform.GridID);
+            var tile = mapGrid.GetTileRef(Owner.Transform.Coordinates.ToVector2i(Owner.EntityManager, _mapManager));
+
+            var solutionFraction = 1 / Math.Floor(_averageLifetime / ExposeDelay);
+
+            // Reagents react with the tile under the smoke
+            foreach (var reagentQuantity in contents.ReagentList.ToArray())
+            {
+                if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                reagent.ReactionTile(tile, reagentQuantity.Quantity * solutionFraction);
+            }
+
+            foreach (var entity in tile.GetEntitiesInTileFast())
+            {
+                // Reagents touch all entities on the tile
+                foreach (var reagentQuantity in contents.ReagentList.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    reagent.ReactionEntity(entity, ReactionMethod.Touch, reagentQuantity.Quantity * solutionFraction);
+                }
```

Comment:
Going through ReagentList thrice seems redundant, you could probably combine all 3 (at the very least these 2).